src/symmetric/tweak_hash_tree.rs

```diff
@@ -1,65 +1,151 @@
 use crate::symmetric::tweak_hash::TweakableHash;
+use rand::Rng;
 use rayon::prelude::*;
 
-/// Hash-Tree based on a tweakable hash function
+/// A single layer of a sparse Hash-Tree
+/// based on tweakable hash function
+struct HashTreeLayer<TH: TweakableHash> {
+    start_index: usize,
+    nodes: Vec<TH::Domain>,
+}
+
+/// Sparse Hash-Tree based on a tweakable hash function.
 /// We consider hash trees in which each leaf is first
 /// hashed individually.
+///
+/// The tree can be sparse in the following sense:
+/// There is a contiguous range of leafs that exist,
+/// and the tree is built on top of that.
+///
+/// For instance, we may consider a tree of depth 32,
+/// but only 2^{26} leafs really exist.
 pub struct HashTree<TH: TweakableHash> {
+    /// Depth of the tree. The tree can have at most
+    /// 1 << depth many leafs. It has depth + 1 many layers
+    depth: usize,
     /// Layers of the hash tree, starting with the
     /// bottom layer. The leafs are not included: the
     /// bottom layer is the list of hashes of all leafs
-    layers: Vec<Vec<TH::Domain>>,
+    layers: Vec<HashTreeLayer<TH>>,
+}
+
+/// Opening in a hash-tree: a co-path, without the leaf
+pub struct HashTreeOpening<TH: TweakableHash> {
+    /// The co-path needed to verify
+    /// If the tree has depth h, i.e, 2^h leafs
+    /// the co-path should have size D
+    co_path: Vec<TH::Domain>,
+}
+
+/// Helper function. Computes a padded layer from the meaningful entries of the layer.
+/// These meaningful entries are assumed to range from start_index to end_index (both inclusive).
+fn get_padded_layer<R: Rng, TH: TweakableHash>(
+    rng: &mut R,
+    nodes: Vec<TH::Domain>,
+    start_index: usize,
+    end_index: usize,
+) -> HashTreeLayer<TH> {
```

Comment:
I think we could simplify this a bit:

```suggestion
fn get_padded_layer<R: Rng, TH: TweakableHash>(
    rng: &mut R,
    nodes: Vec<TH::Domain>,
    start_index: usize,
) -> HashTreeLayer<TH> {
```

Because the `end_index` must be `start_index + nodes.len() - 1`